src/libsystemd-network/sd-ndisc.c

```diff
@@ -500,33 +509,133 @@ _public_ sd_radv *sd_radv_unref(sd_radv *ra) {
                 sd_radv_prefix_unref(p);
         }
 
+        radv_reset(ra);
+
         sd_radv_detach_event(ra);
         return mfree(ra);
 }
 
+static int radv_send(sd_radv *ra, struct in6_addr *dst,
+                     uint32_t router_lifetime) {
+
+        return -ENOSYS;
+}
+
+static usec_t radv_compute_timeout(usec_t min, usec_t max) {
+
+        return MIN(min, max) + (random_u32() % (MAX(min, max) - MIN(min, max)));
+}
+
+static int radv_timeout(sd_event_source *s, uint64_t usec, void *userdata) {
+        int r;
+        sd_radv *ra = userdata;
+        usec_t min_timeout = SD_RADV_DEFAULT_MIN_TIMEOUT;
+        usec_t max_timeout = SD_RADV_DEFAULT_MAX_TIMEOUT;
+        usec_t time_now, timeout;
+        char time_string[FORMAT_TIMESPAN_MAX];
+
+        assert(s);
+        assert(ra);
+        assert(ra->event);
+
+        ra->timeout_event_source = sd_event_source_unref(ra->timeout_event_source);
+
+        r = sd_event_now(ra->event, clock_boottime_or_monotonic(), &time_now);
+        if (r < 0)
+                goto fail;
+
+        if (radv_send(ra, NULL, ra->lifetime) < 0) {
+                log_ndisc("Unable to send Router Advertisment");
+        }
+
+        /* RFC 4861, Section 6.2.4, sending initial Router Advertisments */
+        if (ra->ra_sent < SD_RADV_MAX_INITIAL_RTR_ADVERTISEMENTS) {
+                max_timeout = SD_RADV_MAX_INITIAL_RTR_ADVERT_INTERVAL;
+                min_timeout = SD_RADV_MAX_INITIAL_RTR_ADVERT_INTERVAL / 3;
+        }
+
+        timeout = radv_compute_timeout(min_timeout, max_timeout);
+
+        log_ndisc("Next Router Advertisment in %s",
+                  format_timespan(time_string, FORMAT_TIMESPAN_MAX,
+                                  timeout, USEC_PER_SEC));
+
+        r = sd_event_add_time(ra->event, &ra->timeout_event_source,
+                              clock_boottime_or_monotonic(),
+                              time_now + timeout, MSEC_PER_SEC,
+                              radv_timeout, ra);
+        if (r < 0)
+                goto fail;
+
+        r = sd_event_source_set_priority(ra->timeout_event_source,
+                                         ra->event_priority);
+        if (r < 0)
+                goto fail;
+
+        r = sd_event_source_set_description(ra->timeout_event_source,
+                                            "radv-timeout");
+        if (r < 0)
+                goto fail;
+
+        ra->ra_sent++;
+
+fail:
+        if (r < 0)
+                sd_radv_stop(ra);
+
+        return 0;
+}
+
 _public_ int sd_radv_stop(sd_radv *ra) {
         assert_return(ra, -EINVAL);
 
         log_ndisc("Stopping IPv6 Router Advertisment daemon");
 
+        /* RFC 4861, Section 6.2.5, send at least one Router Advertisment
+           with zero lifetime  */
+        if (radv_send(ra, NULL, 0) < 0)
```

Comment:
The fact that the RFC requires just one goodbye advertisement makes things easy. If there were multiple ones required, we'd need to have timeouts and such. Nice.